velox/type/Decimal.h

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <folly/dynamic.h>
+#include <iomanip>
+#include <sstream>
+#include <string>
+#include <type_traits>
+#include "velox/common/base/Exceptions.h"
+#include "velox/type/StringView.h"
+
+namespace facebook::velox {
+
+#define PRECISION(typmod) ((uint8_t)(typmod >> 8))
+#define SCALE(typmod) ((uint8_t)(typmod))
+#define TYPMOD(X, Y) ((X << 8) | Y)
+#define NUM_DECIMAL_PARAMETERS 2
+
+using int128_t = __int128_t;
+static constexpr uint8_t kMaxPrecisionInt128 = 38;
+static constexpr uint8_t kDefaultScale = 0;
+static constexpr uint8_t kDefaultPrecision = kMaxPrecisionInt128;
+
+/*
+ * This class defines the Velox DECIMAL type support to store
+ * fixed-point rational numbers.
+ */
+class LongDecimal {
+ public:
+  inline const uint8_t getPrecision() const {
+    return precision_;
+  }
+
+  inline const uint8_t getScale() const {
+    return scale_;
+  }
+
+  inline int128_t getUnscaledValue() const {
+    return unscaledValue_;
+  }
+
+  inline void setUnscaledValue(const int128_t& value) {
+    unscaledValue_ = value;
+  }
+
+  operator StringView() const {VELOX_NYI()}
+
+  std::string toString() const;
+
+  operator std::string() const {
+    return toString();
+  }
+
+  bool operator==(const LongDecimal& other) const {
+    return (
+        this->unscaledValue_ == other.getUnscaledValue() &&
+        this->precision_ == other.getPrecision() &&
+        this->scale_ == other.getScale());
+  }
+  bool operator!=(const LongDecimal& other) const {
+    VELOX_NYI();
+  }
+
+  bool operator<(const LongDecimal& other) const {
+    VELOX_NYI();
+  }
+
+  bool operator<=(const LongDecimal& other) const {
+    VELOX_NYI();
+  }
+
+  bool operator>(const LongDecimal& other) const {
+    VELOX_NYI();
+  }
+
+  LongDecimal(
+      int128_t value,
+      uint8_t precision = kDefaultPrecision,
+      uint8_t scale = kDefaultScale)
+      : unscaledValue_(value), precision_(precision), scale_(scale) {}
+  LongDecimal() = default;
+
+ private:
+  int128_t unscaledValue_; // The actual unscaled value with
+                           // max precision 38.
+  uint8_t precision_ = kDefaultPrecision; // The number of digits in unscaled
```

Comment:
Shouldn't these be obtained from the type and not repeated with every value ?

Reply:
We have removed these classes for now.